aggregator: merge x-kubernetes-group-version-kind slices

sttts opened the pull request:
Definitions are supposed to be reused independently from the assign GVKs in the x-kubernetes-group-version-kind field. Different x-kubernetes-group-version-kind can happen if e.g. shared meta/v1 structs like `Status` or `GetOptions` appear in different API groups.

Without this PR, the merging algorithm will create `n` copies of the shared types, which are many in case of kube-apiserver, and even more if there are many CRDs.

Prerequisite for https://github.com/kubernetes/kubernetes/pull/74133.

sttts commented:
/assign @roycaihw @mbohlool

mbohlool commented on pkg/aggregator/aggregator.go:
```diff
@@ -257,3 +267,72 @@ func mergeSpecs(dest, source *spec.Swagger, renameModelConflicts, ignorePathConf
+		if seen[k] {
+			continue
+		}
+		ret = append(ret, x)
```
nit: I would return true if this line at least ran once, i.e. we actually appended anything. this can also be done by checking the length of ret.

sttts replied:
good idea. Fixed.

mbohlool commented on pkg/aggregator/aggregator.go:
```diff
@@ -257,3 +267,72 @@ func mergeSpecs(dest, source *spec.Swagger, renameModelConflicts, ignorePathConf
 	return nil
 }
+
+// deepEqualDefinitionsModuloGVKs compares a and b, but ignores the x-kubernetes-group-version-kind extension.
```
can this temporary modification changes the order of fields? that could be ignorable, but I prefer not to mutate the objects we promised we won't even temporarily. what about checking deepEqual on spec.Schema.SchemaProps and spec.Schema.SwaggerSchemaProps and manually on spec.Schema.Extensions?

sttts replied:
actually we can just set the whole map to nil, do comparison and then compare the old values independently. No modification necessary.

sttts replied:
fixed

mbohlool commented on pkg/aggregator/aggregator.go:
```diff
@@ -257,3 +267,72 @@ func mergeSpecs(dest, source *spec.Swagger, renameModelConflicts, ignorePathConf
+// deepEqualDefinitionsModuloGVKs compares a and b, but ignores the x-kubernetes-group-version-kind extension.
+// It will mutate a temporarily, but not b.
+func deepEqualDefinitionsModuloGVKs(a, b *spec.Schema) bool {
+	old, oldFound := a.ExtraProps["x-kubernetes-group-version-kind"]
```
ExtraProps is not serialized and not what we use to store GVK. it is spec.Schema.VendorExtensible

sttts replied:
good point. Fixed.

sttts commented:
@mbohlool addressed all your comments. Thanks for the review.

roycaihw commented:
talked with @mbohlool before. still trying to understand the use case here. maybe an example would help  explain (what different gvk extensions will be assigned to the same Go type definition (for delegated API servers))?

> Definitions are supposed to be reused independently from the assign GVKs in the x-kubernetes-group-version-kind field

the statement I totally agree. LGTM

mbohlool commented:
The code looks good now. An example usecase of this would be nice. @sttts  @roycaihw

roycaihw commented on pkg/aggregator/aggregator.go:
```diff
@@ -257,3 +267,71 @@ func mergeSpecs(dest, source *spec.Swagger, renameModelConflicts, ignorePathConf
+		if s1.Extensions != nil {
+			shallowCopy := *s1
+			s1 = &shallowCopy
+			s1.Extensions = nil
```
it's okay to do so now, but we need to keep in mind that this function is actually `Modulo"Extensions"`. New extension key could be dropped silently during merging if we add any in future

sttts replied:
fixed. And added tests.

sttts commented:
@roycaihw @mbohlool the example is already in the description: shared meta/v1 structs like Status or GetOptions appear in different API groups.

sttts commented on pkg/aggregator/aggregator.go:
```diff
@@ -207,7 +207,12 @@ func mergeSpecs(dest, source *spec.Swagger, renameModelConflicts, ignorePathConf
+					if gvks, found, err := mergedGVKs(&v2, &v); err != nil {
+						return err
+					} else if found {
+						v2.Extensions["x-kubernetes-group-version-kind"] = gvks
```
it mutates the `dest` which is intentional and correct.

sttts commented:
@roycaihw this is pre-requisite for the server-side-apply wiring for CRDs. I addressed your comment, including new unit tests for the equality-module-gvks. Ptal.

sttts commented:
/remove-lifecycle rotten

sttts commented:
/approve

p0lyn0mial commented on pkg/aggregator/aggregator.go:
```diff
@@ -257,3 +269,96 @@ func mergeSpecs(dest, source *spec.Swagger, renameModelConflicts, ignorePathConf
+		return gvk2, found2, nil
+	}
+	if !found2 {
+		return gvk1, found1, nil
```
nit: shouldn't this be `gvk1, false, nil` ? If we haven't found `s2` we don't change `s1` thus this is noop

sttts replied:
yep, fixed.

p0lyn0mial commented on pkg/aggregator/aggregator.go:
```diff
@@ -257,3 +269,96 @@ func mergeSpecs(dest, source *spec.Swagger, renameModelConflicts, ignorePathConf
+
+	ret := make([]interface{}, len(slice1), len(slice1)+len(slice2))
+	copy(ret, slice1)
+	seen := make(map[string]bool, len(slice1)+len(slice2))
```
nit: I think that `seen` could be only `len(slice1)`.

sttts replied:
fixed

sttts commented:
@p0lyn0mial addressed your comments.

p0lyn0mial commented:
/lgtm
